### ISSUE TITLE
Combining options and positional arguments?

### ISSUE THREAD

**Issue by dacto** [Gabriel439/Haskell-Optparse-Generic-Library#65]
I'd like to have required options (--foo, --bar) and a positional argument that consumes all remaining things on the command line into a list.

#### Sample Program
```haskell
import Options.Generic

data RequiredOptions = RequiredOptions
    { foo :: FilePath <?> "specify foo option filepath"
    , bar :: Text     <?> "specify bar option text"
    }
    deriving (Generic, Show)

instance ParseRecord RequiredOptions

data PositionalArgs = PositionalArgs ([Text] <?> "all other command line arguments")
    deriving (Generic, Show)

instance ParseRecord PositionalArgs

data Options = Options RequiredOptions PositionalArgs
    deriving (Show)

instance ParseRecord Options where
    parseRecord = Options <$> parseRecord <*> parseRecord

main = do
    x <- getRecord "Test program"
    print (x :: Options)
```
Output:
```
Test program

Usage: test --foo STRING --bar TEXT [TEXT...]

Available options:
  -h,--help                Show this help text
  --foo STRING             specify foo option filepath
  --bar TEXT               specify bar option text
  -h,--help                Show this help text
  TEXT...                  all other things
```

#### Usage example
```
> test --foo "/some/file/path" --bar "word" --extra-arg1 "val1" --extra-arg2 "val2" "extra-pos-arg"
Invalid option `--extra-arg1'

Usage: test --foo STRING --bar TEXT [TEXT...]

> test --foo "/some/file/path" --bar "word" -- --extra-arg1 "val1" --extra-arg2 "val2" "extra-pos-arg"
Options (RequiredOptions {foo = Helpful {unHelpful = "/some/file/path"}, bar = Helpful {unHelpful = "word"}}) (PositionalArgs (Helpful {unHelpful = ["--extra-arg1","val1","--extra-arg2","val2","extra-pos-arg"]}))
```

Given the sample program and output, it appears I am _successfully_ combining options and positional arguments, it prompts the following questions:
1. Why are there duplicate `-h,--help ..` lines in the usage message? How to avoid this?
2. Is it possible to avoid requiring a `--` to separate the required options from the positional arguments?
  In other words, is it possible to avoid requiring a `--` if all that remains are unknown arguments? Can all unknown arguments become the list of positional arguments with needing a `--`?
3. Is there a better way to combine options and positional arguments? 

`optparse-generic version`: `1.4.0`

**Comment by Gabriella439** [comment]
@dacto: The package wasn't really designed for this use case, so the fact that it works is more of an accident 🙂 

I probably won't have time to fix this myself since it wasn't something I originally intended to support, but I'd accept a pull request if somebody else were more inclined

**Comment by foxbenjaminfox** [comment]
I came up with what I think is a neat way of combining options and positional arguments:

```haskell
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Options.Generic

data Options = Options {
  foo :: Int,
  bar :: Double,
  baz :: NoLabel String,
  quux :: NoLabel Int
} deriving (Generic, Show)

instance ParseRecord Options

newtype NoLabel a = NoLabel a
  deriving (Generic, Show)

instance ParseFields a => ParseRecord (NoLabel a)
instance ParseFields a => ParseFields (NoLabel a) where
  parseFields msg _ _ def = fmap NoLabel (parseFields msg Nothing Nothing def)

main :: IO ()
main = do
    x <- getRecord "Test program"
    print (x :: Options)
```

This lets you have a static number of typed positional parameters in addition to options. It uses newtype wrapping kind of like what `<?>` and `<!>` do, and in theory if included in the `optparse-generic` package itself could be unwrapped by the `Unwrap` type family along with `<?>` and `<!>`.

I'd put this forward as an answer to question 3: this seems like a better way of doing it (even if you do need to unwrap the `NoLabel` newtype yourself.) It doesn't have issues 1 and 2 — ``--help` shows up only once in the usage message, and you can mix positional arguments and options however you please:

```
> test hello 1 --foo 2 --bar 3.14
Options {foo = 2, bar = 3.14, baz = NoLabel  "hello", quux = NoLabel 1}

> test --foo 2 --bar 3.14 hello 1
Options {foo = 2, bar = 3.14, baz = NoLabel "hello", quux = NoLabel 1}

> test --foo 2 hello --bar 3.14 1
Options {foo = 2, bar = 3.14, baz = NoLabel "hello", quux = NoLabel 1}
```

**Comment by Gabriella439** [comment]
@foxbenjaminfox: Yeah, that seems like a good approach!

**Comment by dacto** [comment]
@foxbenjaminfox this is great! Very awesome! Thank you :smile:

Note that this approach also does eliminate the need for `--` for separating known options from following unknown ones. Where the unknown ones should get used as the positional arguments.

Example:
```
let o = option, p = positional in:

>  test --known1 k2  --known2 k2  --unknown1  u1  "another thing"
ideal:  [   o1    ]  [   o2    ]  [   p1   ] [p2] [    p3       ]
actual:  error about --unknown1 being invalid option.

>  test --known1 k2  --known2 k2  -- --unknown1  u1  "another thing"
actual: [   o1    ]  [   o2    ]     [   p1   ] [p2] [    p3       ]
```
`optparse-applicative` has a [`forwardOptions`](https://www.stackage.org/haddock/nightly-2020-06-26/optparse-applicative-0.15.1.0/Options-Applicative.html#v:forwardOptions) info modifier which some how could be used here.
Maybe I can incorporate that into your solution.

Otherwise, I'll make do with `--` :smiley:

**Comment by raboof** [comment]
This seems like a great utility - would it make sense to include the `NoLabel` type in the library?
